In [32]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [33]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [34]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [35]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [36]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()

In [37]:
# Create a DataFrame with the scaled data
scaled_data_df = pd.DataFrame(
    scaler.fit_transform(df_market_data),
    columns = df_market_data.columns)


# Copy the crypto names from the original DataFrame
scaled_data_df["coin_id"] = df_market_data.index


# Set the coin_id column as index   
scaled_data_df = scaled_data_df.set_index("coin_id")



# Display the scaled DataFrame
scaled_data_df.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [38]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 11))



In [39]:
# Create an empty list to store the inertia values
inertia = []


# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for k in k_values:
    model = KMeans(n_clusters=k, random_state=0)
    model.fit(scaled_data_df)
    inertia.append(model.inertia_)


c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Window

In [40]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k": k_values,
    "inertia": inertia
}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)


In [41]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
scaled_elbow = df_elbow.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve for Optimal k',
    xticks=k,
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    width=800,
    height=400
)
# Display the elbow plot
scaled_elbow


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 
4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [42]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [43]:
# Fit the K-Means model using the scaled DataFrame
model.fit(scaled_data_df)

c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [44]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
scaled_data_df["class"] = model.predict(scaled_data_df)

# Print the resulting array of cluster values.
scaled_data_df["class"].value_counts() 

class
0    26
2    13
1     1
3     1
Name: count, dtype: int64

In [45]:
# Create a copy of the scaled DataFrame
df_scaled_cluster = scaled_data_df.copy()

In [46]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_scaled_cluster["class"] = model.labels_

# Display the copy of the scaled DataFrame
df_scaled_cluster.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  class  
coin_id                                             
bitcoin                           -0.251637      2  
ethereum                          -0.199352      2  
tether                            -0.282061      0  
ripple                            -0.295546      0  
bitcoin-cash                      -0.270317      2  
binancecoin                       -0.225533      2  
chainlink                         -0.018284      2  
cardano                           -0.155428      2  
litecoin                          -0.292351      2  
bitcoin-cash-sv                   -0.206029      0

In [47]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
cluster_v1 = df_scaled_cluster.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="class",
    hover_cols=["coin_id"],
    title="Cryptocurrency Clusters based on Price Changes",
    xlabel="24h Price Change (%)",
    ylabel="7d Price Change (%)",
    width=800,
    height=400
)
cluster_v1

:NdOverlay   [class]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [48]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [49]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
reduced_data_pca = pca.fit_transform(df_scaled_cluster)
# View the scaled PCA data
reduced_data_pca[:5]  # Display the first 5 rows of the reduced data
# Create a DataFrame with the PCA data and assign the cluster labels

array([[ 1.19408229, -0.90207402,  0.58533763],
       [ 1.00970605, -0.66358438,  1.13695032],
       [-0.72253591, -0.30763241, -0.7218133 ],
       [-0.74826646, -0.33237895, -0.55847869],
       [ 2.25853853, -1.82696614,  1.37816649]])

In [50]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_
#Add these variance ratios together to see how much information is retained
total_variance = sum(pca.explained_variance_ratio_)
total_variance

0.8867743104975921

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 88.68%

In [51]:
# Create a new DataFrame with the PCA data.
pca_df = pd.DataFrame(
    reduced_data_pca,
    columns=["PCA1", "PCA2", "PCA3"],
    index=scaled_data_df.index
)




# Copy the crypto names from the original scaled DataFrame
pca_df["coin id"] = scaled_data_df.index

# Set the coin_id column as index
pca_df = pca_df.set_index("coin id")

# Display the scaled PCA DataFrame
pca_df.head(10)

PCA1      PCA2      PCA3
coin id                                      
bitcoin          1.194082 -0.902074  0.585338
ethereum         1.009706 -0.663584  1.136950
tether          -0.722536 -0.307632 -0.721813
ripple          -0.748266 -0.332379 -0.558479
bitcoin-cash     2.258539 -1.826966  1.378166
binancecoin      1.719188 -0.993613  0.663561
chainlink        1.396401 -0.721542  2.731341
cardano          1.359187 -0.651114  1.482162
litecoin         0.869062 -0.828724  0.856779
bitcoin-cash-sv -0.763186 -0.619275 -0.315895

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [52]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [53]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list  
for k in range(1, 11):
    model = KMeans(n_clusters=k, random_state=0)
    model.fit(pca_df[["PCA1", "PCA2", "PCA3"]])
    inertia.append(model.inertia_)



c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Window

In [54]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    "k": list(range(1, 11)),
    "inertia": inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)


In [55]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.   
pca_elbow =df_elbow_pca.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve for Optimal k (PCA)',
    xticks=list(range(1, 11)),
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    width=800,
    height=400
)
pca_elbow

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** no it is not different

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [56]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)


In [57]:
# Fit the K-Means model using the PCA data
model.fit(pca_df)

c:\Users\the_a\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [58]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_df["class"] = model.predict(pca_df)


# Print the resulting array of cluster values.
pca_df["class"].value_counts()
# pca_df.head(10)  # Display the first 10 rows of the PCA DataFrame with clusters


class
0    26
1    13
2     1
3     1
Name: count, dtype: int64

In [59]:
# Create a copy of the scaled PCA DataFrame
df_scaled_cluster = pca_df.copy()  # Make a copy of the PCA DataFrame



# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_scaled_cluster["class"] = model.labels_

# Display the copy of the scaled PCA DataFrame
df_scaled_cluster.head(10)  # Display the first 10 rows of the PCA DataFrame with clusters

PCA1      PCA2      PCA3  class
coin id                                             
bitcoin          1.194082 -0.902074  0.585338      1
ethereum         1.009706 -0.663584  1.136950      1
tether          -0.722536 -0.307632 -0.721813      0
ripple          -0.748266 -0.332379 -0.558479      0
bitcoin-cash     2.258539 -1.826966  1.378166      1
binancecoin      1.719188 -0.993613  0.663561      1
chainlink        1.396401 -0.721542  2.731341      1
cardano          1.359187 -0.651114  1.482162      1
litecoin         0.869062 -0.828724  0.856779      1
bitcoin-cash-sv -0.763186 -0.619275 -0.315895      0

In [60]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
cluster_v2 = df_scaled_cluster.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="class",
    hover_cols=["coin id"],
    title="Cryptocurrency Clusters based on PCA",
    xlabel="PCA 1",
    ylabel="PCA 2",
    width=800,
    height=400
)
# Display the PCA cluster plot
cluster_v2

:NdOverlay   [class]
   :Scatter   [PCA1]   (PCA2,coin id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [61]:
# Composite plot to contrast the Elbow curves
scaled_elbow + pca_elbow

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [62]:
# Composite plot to contrast the clusters
cluster_v1 + cluster_v2

:Layout
   .NdOverlay.I  :NdOverlay   [class]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [class]
      :Scatter   [PCA1]   (PCA2,coin id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**  The impact of using fewer features to cluster the data using K-Means, is that we see better more conscise grouping. There is way more seperation from the centroid and cleared grouping of clusters.